week2/lc934_shortestBridge.js

```diff
@@ -0,0 +1,67 @@
+// Runtime: 120 ms, faster than 61.57% 
+// Memory Usage: 42 MB, less than 93.75%
+
+/**
+ * @param {number[][]} A
+ * @return {number}
+ */
+var shortestBridge = function(A) {
+    let queue = [];
+    let dx = [0, 0, 1, -1];
+    let dy = [1, -1, 0, 0];
+    let mark = false;
+    for(let i = 0; i < A.length; i++) {
+        if(mark) break;
+        for(let j = 0; j < A[0].length; j++) {
+            if(A[i][j] === 1) {
+                dfs(A, i, j, queue);
+                mark = true;
+                break;
+            }
+        }
+    }    
+    
+    let ret = 0;
+    
+    while (queue.length > 0) {
+        let size = queue.length;
+        for (let i = 0; i < size; i++) {
+            let curr = queue.shift();

```

Comment:
shift has bad performance. ~ref:
https://stackoverflow.com/questions/6501160/why-is-pop-faster-than-shift#:~:text=shift()%20has%20to%20re,do%20not%20move%3B%20simply%20the%20.&text=shift()%20removes%20the%20first%20element%20in%20the%20array.

Reply:
Yes, but here in my bfs solution, the pop() function can't meet my requirement. The element gets removed has to be at the start of the queue(array), not at the end of the queue. I keep removing the previous level nodes at each iteration.

Reply:
We sure can use the way you wrote, in that case the pop() function is enough, just different styles of bfs.

Reply:
let size = queue.length;
        for (let i = 0; i < size; i++) {...}

You mean this is not a level searching? In this case, what do you have in the queue? 
From the codes, you calculate the size of the queue, then iteratively pick every point in the queue. They are in the same level.

The real issue here is in line40. You keep adding new points to this queue. This seems correct, but you may need to create a temp queue to store those `next-level` points. And after searching every point in the queue, copy all points in this temp queue to the queue. How do you think?

Reply:
Yes, every time we re-enter a while loop (line 26), it's one level searching. Each iteration, after the for loop (line 28) it removes the current level points out of queue from the head, then adds all points next level from the tail. So, the next 
while iteration the new queue will only include nodes the level of nodes just added in the last iteration. 

I guess you mean we use a team queue only store the level of nodes we just added, then each time we reassign the queue with temp queue, in that way we can still use pop() or shift() (of course in this case pop() is better). But each time we create a new temp queue memory space, in my solution I use same queue all the time, but in javascript, not like java has queue collection, we have to shift the whole array to left, it's consuming more elements*(number of nodes shift out) time complexity. In your suggestion, we only consume more number of level of nodes space complexity.

Reply:
Use queue.shift() version, it consumes (number of elements in queue now)*(number of element remove from head) time complexity,
User queue.pop() with teamQueue, it consumes (number of elements are going to add in this level) space complexity. Seem like the second one is more efficient.

Reply:
You even don't need the pop or shift:

`for (let i = 0; i < size; i++) {
    let curr = queue[I];
}`

It would be faster